File: benchmark/synduce/expressions/max_subexpr_sum.f
Inductive ArithExpr = int Int | plus {ArithExpr, ArithExpr} | minus {ArithExpr, ArithExpr};
Inductive PsumExpr = nint Int | nplus {PsumExpr, PsumExpr} | nminus {PsumExpr, PsumExpr};

max = \a: Int. \b: Int. if < a b then b else a;

repr = fix (
  \f: PsumExpr -> Compress ArithExpr. \e: PsumExpr.
  match e with
    nint i -> int i
  | nplus {e1, e2} -> plus {f e1, f e2}
  | nminus {e1, e2} -> minus {f e1, f e2}
  end
);

spec = \e: ArithExpr. (fix (
  \f: ArithExpr -> {Int, Int}. \e: ArithExpr.
  match e with
    int i -> {i,i}
  | plus {e1, e2} ->
    let r1 = f e1 in
    let r2 = f e2 in
    let sum = + r1.1 r2.1 in
    let m = max sum (max r1.2 r2.2) in
    {sum, m}
  | minus {e1, e2} ->
    let r1 = f e1 in
    let r2 = f e2 in
    let sum = - r1.1 r2.1 in
    let m = max sum (max r1.2 r2.2) in
    {sum, m}
  end
) e).2;

main = \e: PsumExpr. spec (repr e);

File: ml-benchmark/synduce/expressions/max_subexpr_sum.ml
type arith_expr =
  | Int of int
  | Plus of arith_expr * arith_expr
  | Minus of arith_expr * arith_expr

type psum_expr =
  | NInt of int
  | NPlus of int * psum_expr * psum_expr
  | NMinus of int * psum_expr * psum_expr

let max a b = if a > b then a else b

val repr: psum_expr -> arith_expr compress
let rec repr = function
  | NInt i -> Int i
  | NPlus (s, a, b) -> Plus (repr a, repr b)
  | NMinus (s, a, b) -> Minus (repr a, repr b)

let rec sum = function
  | NInt i -> i
  | NPlus (s, a, b) -> sum a + sum b
  | NMinus (s, a, b) -> sum a - sum b

let rec is_memo_psum = function
  | NInt i -> true
  | NPlus (s, a, b) -> s >= sum a + sum b && is_memo_psum a && is_memo_psum b
  | NMinus (s, a, b) -> s >= sum a - sum b && is_memo_psum a && is_memo_psum b

let rec spec = function
  | Int i -> (i, i)
  | Plus (a, b) ->
    (match spec a with (asum, am) -> 
      match spec b with (bsum, bm) -> 
        (asum + bsum, max (asum + bsum) (max am bm)))
  | Minus (a, b) ->
    (match spec a with (asum, am) -> 
      match spec b with (bsum, bm) -> 
        (asum - bsum, max (asum - bsum) (max am bm)))

let prog e = spec (repr e)